File: backend/src/domain/task_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional
from .task import Task, TaskCreate, TaskUpdate

class TaskRepository(ABC):
    @abstractmethod
    def create(self, task: TaskCreate) -> Task:
        pass

    @abstractmethod
    def get_all(self) -> List[Task]:
        pass

    @abstractmethod
    def get_by_id(self, task_id: int) -> Optional[Task]:
        pass

    @abstractmethod
    def update(self, task_id: int, task_update: TaskUpdate) -> Optional[Task]:
        pass

    @abstractmethod
    def delete(self, task_id: int) -> bool:
        pass

File: internal/ports/task_repository.go
// internal/ports/task_repository.go
package ports

import (
	"context"
	"devops_console/internal/domain/entities"
	"devops_console/internal/domain/workspace"
)

type TaskFilters struct {
	workspaceID string
	taskType    string
	subjects    []entities.Subject
}

type TaskUpdate struct {
	Name        string
	Description string
	Config      entities.TaskConfig
	workspace.Workspace
	entities.TaskType
	Approvals []entities.Approval
	Triggers  []entities.Trigger
}

// TaskRepository define las operaciones de persistencia para las tareas.
type TaskRepository interface {
	Create(ctx context.Context, task *entities.DevOpsTask) error
	GetByID(ctx context.Context, taskID string) (entities.DevOpsTask, error)
	Update(ctx context.Context, task *entities.DevOpsTask) error
	Delete(ctx context.Context, taskID string) error
	GetAll(ctx context.Context, filters TaskFilters) ([]entities.DevOpsTask, error)
	GetByExecutionID(ctx context.Context, executionID string) (entities.DevOpsTask, error)
}
